third_party/move/move-vm/runtime/src/runtime_ref_checks.rs

```diff
@@ -1713,6 +1744,10 @@ impl RefCheckState {
         num_locals: usize,
         mask: ClosureMask,
     ) -> PartialVMResult<()> {
+        // when calling a borrow field mut api, we directly check mutably borrowing the field at offset
```

Comment:
Why do we want to do this directly? Does the existing ref checker semantics (by going over the bytecode in the function body) not handle this correctly? Are you implementing an optimization?

Also, we should definitely not be adding arbitrary (undocumented) invariants like caller sets some flag, and is cleared by callee in some cases: there is a clean way to do this if necessary (my updated PR passes the loaded function, so you should be able to get info from there).

Reply:
I think it is similar to the logic in bytecode verifier: the call to borrow mut API takes the mutable reference as input, if we don't special case here, the check will assume there may be destructive write via this mut ref.

Reply:
I think we need to clearly document the reasoning (why we need this bypass, ideally with the simplest example, or have a test case and point to it). This will help future refactorings.

For example, if I understand the core reason correctly: instead of having to maintain internal state across different bytecode transitions, we could also just modify the call (if it has appropriate attributes) to skip the destruction write step (which IMO is much more localized and clean, as opposed to having to reason across multiple transitions.

We also need a test case in the runtime ref checks verifier that shows invariant violation before and no invariant violation after.

Reply:
refactored the code to return early in `pre_execution_transition`, `core_call` and `post_execution_transition` when the function is the borrow mut API instead of doing byte code transition inside the code.

Reply:
Yeah, a legitimate use case of why the destructive update is too conservative and how the attribute solves it. I think we are on the same page.

Reply:
Added two masm cases.